File: SWEA/D3/6808. 규영이와 인영이의 카드게임/규영이와 인영이의 카드게임.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main(){
    int T;
    cin >> T;
    for(int t = 1; t <= T; t++){
        int G[9];
        bool used[19] = {false};

        for(int i = 0; i < 9; i++){
            cin >> G[i];
            used[G[i]] = true;
        }

        int I[9];
        int idx = 0;
        for(int i = 1; i <= 18; i++){
            if(!used[i]){
                I[idx] = i;
                idx++;
            }
        }
        
        int win = 0, lose = 0;

        sort(I, I + 9);

        do{
            int g_score = 0, i_score = 0;
            for(int i = 0; i < 9; i++){
                if(G[i] > I[i]) g_score += G[i] + I[i];
                else i_score += G[i] + I[i];
            }
            if(g_score > i_score) win++;
            else if(g_score < i_score) lose++;
        }while(next_permutation(I, I + 9));
        cout << "#" << t << ' ' << win << ' ' << lose << '\n';
    }
    return 0;
}

File: SWEA/D3/6808. 규영이와 인영이의 카드게임/규영이와 인영이의 카드게임.java
import java.util.*;
import java.io.*;

public class Solution {
    static int win, lose;
    static int[] G = new int[9];
    static int[] I = new int[9];
    static boolean[] use = new boolean[19];
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        for(int test_case = 1; test_case <= T; test_case++){
            StringTokenizer st = new StringTokenizer(br.readLine());

            
            boolean[] get = new boolean[19];

            for(int i = 0; i < 9; i++){                     //규영이 카드 채우고
                G[i] = Integer.parseInt(st.nextToken());
                get[G[i]] = true;
            }

            int idx = 0;
            for(int i = 1; i < 19; i++){                     //남는거 인영이 짬처리
                if(!get[i]) {
                    I[idx] = i;
                    idx++;
                }
            }

            win = 0;
            lose = 0;
            Arrays.fill(use, false);

            fx(0, 0, 0);

            System.out.println("#" + test_case + " " + win + " " + lose);
        }
    }
    static void fx(int d, int Gscore, int Iscore){
        if(d == 9){
            if(Gscore > Iscore) win++;
            else lose++;
            return;
        }
        for(int i = 0; i < 9; i++){
            if(!use[i]){
                use[i] = true;
                int sum = G[d] + I[i];
                if(G[d] > I[i]){
                    fx(d + 1, Gscore + sum, Iscore);
                }else {
                    fx(d + 1, Gscore, Iscore + sum);
                }
                use[i] = false;
            }
        }
    }
}
